### PR TITLE
feat(quickstarts/tour): add Cryostat guided tour and various QuickStarts

### PR DESCRIPTION
**Pull request by maxcao13**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] Signed the last commit: `git commit --amend --signoff`
_______________________________________________

Fixes: #902 
Closes: #880

## Description of the change:
Adds a guided tour to Cryostat, similar to what Openshift does and adds 4? new quickstarts. These quickstarts and guide can be found under the (?) help button.

## How to manually test:
Run normally and try each of these components out.
image

### CODE REVIEW

**Comment by maxcao13** [comment]
Should be ready for light review, settings quickstart could use extra work, and dashboard quick start has been started yet.

**Comment by tthvo** [comment]
Looks amazing! I am blown away :D Just a notice that some references to an element is not working (maybe the element is lacking the `data-highlight` props). For example, the event template selector in Automated Rules. We might also have to check collapsed navigation sidecar and actions overflow button in Recording table.

**Comment by andrewazores** [comment]
image

> allow you start recordings on a target JVM 

I think this should be pluralized: "allow you to start recordings on target JVMs based on sets of conditions"

image

Any particular reason for calling out reports here? Reports can also be generated on the Recordings and Dashboards views, after all.

image

> The Events page lists the Event Templates that Cryostat can currently apply to a JFR recording. It also lists Event Types that are available on the target JVM.

"The Events page lists the Event Templates that can be used for creating Flight Recordings. It also details the JFR Event Types that can be recorded within each target JVM."

File: README.md [hunk 1]
```diff
@@ -90,3 +90,45 @@ The extraction tool is [`i18next-parser`](https://www.npmjs.com/package/i18next-
...
 
 To workaround this, specify static values in `i18n.ts` file under any top-level directory below `src/app`. For example, `src/app/Settings/i18n.ts`.
+
+## ADDING QUICKSTARTS
+
+To add a new quickstart, create a new tsx/ts file under `src/app/QuickStarts` with your Quick Start name, like `my-quickstart.tsx`. 
+
+Cryostat's Quick Starts use a markdown extension which allows components to be highlighted using a button within the markdown in the Quick Start content itself. It was taken from OpenShift Console's GitHub repo and modified to fit Cryostat's needs.
+
+The following are taken from patternfly/patternfly-quickstarts GitHub repo.
```
Comment (tthvo) [inline README.md:100]: We probably dont need to mention this. Or maybe add something like `See More` if the content has more details on the PF Quickstart docs.

Comment (tthvo) [reply README.md:100]: And maybe remove this: `It was taken from OpenShift Console's GitHub repo and modified to fit Cryostat's needs.`

File: src/app/BreadcrumbPage/BreadcrumbPage.tsx [hunk 2]
```diff
@@ -52,7 +60,7 @@ export interface BreadcrumbTrail {
 
 export const BreadcrumbPage: React.FC<BreadcrumbPageProps> = (props) => {
   return (
-    <>
+    <PageGroup>
       <PageSection>
         <Breadcrumb>
```
Comment (tthvo) [inline src/app/BreadcrumbPage/BreadcrumbPage.tsx:65]: Wrap this in `PageBreadCrumb`?

See: https://www.patternfly.org/v4/components/page/#group-section

File: src/app/assets/palette.svg
Comment (tthvo) [inline src/app/assets/palette.svg:1]: Is thistaken from Red Hat design files for Cryostat? If not, we might need to take it from there.

File: src/app/app.css [hunk 3]
```diff
@@ -37,6 +37,18 @@ SOFTWARE.
 */
 html, body, #root {
   height: 100%;
+  overflow: hidden;
```
Comment (tthvo) [inline src/app/app.css:40]: Any reason to set overflow as hidden here?

Comment (tthvo) [reply src/app/app.css:40]: Same for `.pf-c-page__sidebar.pf-m-collapsed`.

File: src/app/Joyride/JoyrideProvider.tsx [hunk 4]
```diff
@@ -35,46 +35,54 @@
...
 
-export default withRouter(QuickStarts);
+export const useJoyride = (): {
+  setState: (patch: Partial<JoyrideState> | ((previousState: JoyrideState) => Partial<JoyrideState>)) => void;
+  state: JoyrideState;
+  isNavBarOpen: boolean;
+  setIsNavBarOpen: React.Dispatch<React.SetStateAction<boolean>>;
+} => {
+  const context = React.useContext(JoyrideContext);
+  if (context === undefined) {
```
Comment (tthvo) [inline src/app/Joyride/JoyrideProvider.tsx:84]: This might not ever be reached since if no provider is found for this context, the default value is returned.

File: src/app/Joyride/JoyrideProvider.tsx [hunk 5]
```diff
@@ -35,46 +35,54 @@
...
-  const [allQuickStartStates, setAllQuickStartStates] = useLocalStorage('quickstarts', {});
+const defaultState = {
+  run: false,
+  stepIndex: 0,
+  steps: [] as Step[],
+};
+
+/* eslint-disable @typescript-eslint/no-empty-function */
+export const JoyrideContext = React.createContext({
+  state: defaultState,
```
Comment (tthvo) [inline src/app/Joyride/JoyrideProvider.tsx:56]: Extract to a type/interface?

File: src/app/QuickStarts/QuickStartDrawer.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+import { useTranslation } from 'react-i18next';
+import { distinctUntilChanged } from 'rxjs';
+import GenericQuickStart from './quickstarts/generic-quickstart';
+
+const LINK_LABEL = "[\\d\\w\\s-()$!&']+"; // has extra &' in matcher
+const HIGHLIGHT_ACTIONS = ['highlight']; // use native quickstarts highlight markdown extension
+const SELECTOR_ID = `[\\w-&]+`; // has extra &'
+
+// [linkLabel]{{action id}}
+const HIGHLIGHT_REGEXP = new RegExp(`\\[(${LINK_LABEL})]{{(${HIGHLIGHT_ACTIONS.join('|')}) (${SELECTOR_ID})}}`, 'g');
```
Comment (tthvo) [inline src/app/QuickStarts/QuickStartDrawer.tsx:61]: Escape the second square bracket `]`?

File: src/app/QuickStarts/QuickStartDrawer.tsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+  React.useEffect(() => {
+    addSubscription(
+      context.settings
+        .featureLevel()
+        .pipe(distinctUntilChanged())
+        .subscribe((level) => {
+          if (level === FeatureLevel.DEVELOPMENT) {
+            if (!allQuickStarts.includes(GenericQuickStart)) {
+              allQuickStarts.push(GenericQuickStart);
```
Comment (tthvo) [inline src/app/QuickStarts/QuickStartDrawer.tsx:81]: Ideally, exported variables should be read-only.

Above, it is actually quite unsafe since somehow this array is updated before quickstart finishes initiates. Try wrapping this callback in `setTimeout` with 500ms, set to DEV Mode, notice the generic quickstart is not showing up.

Is there anyway we can dynamically update the total list of quickstarts?

Comment (tthvo) [reply src/app/QuickStarts/QuickStartDrawer.tsx:81]: If possible, we can add a metadata to the quickstart object to hold the featureLevel, then call `useFeatureLevel` and use the currect level to filter out quickstarts.

File: src/app/Joyride/CryostatJoyride.tsx [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+              <h1>
+                Welcome to <strong>Cryostat</strong>!
+              </h1>
+            </div>
+          ),
+          target: 'body',
+          disableBeacon: true,
+        },
+        {
+          content: <p>Open the navigation bar!</p>,
```
Comment (tthvo) [inline src/app/Joyride/CryostatJoyride.tsx:84]: Not sure if I did some thing wrong but I am seeing this:

Screenshot from 2023-03-28 01-29-09

File: src/app/Joyride/CustomTooltip.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+  CardTitle,
+  Split,
+  SplitItem,
+  Text,
+  TextContent,
+} from '@patternfly/react-core';
+import React from 'react';
+import { TooltipRenderProps } from 'react-joyride';
+
+const CustomTooltip: React.FC<TooltipRenderProps> = ({
```
Comment (tthvo) [inline src/app/Joyride/CustomTooltip.tsx:55]: Lets call it `JoyRideTooltip`?

File: src/app/AppLayout/AppLayout.tsx [hunk 10]
```diff
@@ -326,38 +345,36 @@ const AppLayout: React.FC<AppLayoutProps> = ({ children }) => {
...
+  const handleOpenGuidedTour = React.useCallback(() => {
+    setJoyState({ run: true });
+  }, [setJoyState]);
+
+  const helpItems = React.useMemo(() => {
+    return [
+      <ApplicationLauncherItem
+        key={'Quickstarts'}
+        component={<NavLink to="/quickstarts">{t('AppLayout.APP_LAUNCHER.QUICKSTARTS')}</NavLink>}
+      ></ApplicationLauncherItem>,
```
Comment (tthvo) [inline src/app/AppLayout/AppLayout.tsx:357]: Use `< Component />` since it has no children?

File: src/app/AppLayout/AppLayout.tsx [hunk 11]
```diff
@@ -326,38 +345,36 @@ const AppLayout: React.FC<AppLayoutProps> = ({ children }) => {
...
+      <ApplicationLauncherItem key={'Documentation'} onClick={handleOpenDocumentation}>
+        <span>{t('AppLayout.APP_LAUNCHER.DOCUMENTATION')}</span>
+        <Icon isInline size="lg" iconSize="sm" style={{ marginLeft: 'auto', paddingLeft: '1ch' }}>
```
Comment (tthvo) [inline src/app/AppLayout/AppLayout.tsx:360]: Nice css tricks btw! Been looking for how to align the icons!

**Review by tthvo — COMMENTED** [review]
Looks awesome! Any other planned quickstarts for this PR? For topology, probably better to be in a follow-up.

File: src/app/Joyride/CryostatJoyride.tsx (hunk 8 above)
Comment (maxcao13) [reply src/app/Joyride/CryostatJoyride.tsx:84]: Yes, it should... is there a way to replicate this?

File: src/app/assets/palette.svg
Comment (maxcao13) [reply src/app/assets/palette.svg:1]: Hmm... no I just tried color picking the Cryostat logo because I wanted to fit the cryostat-tour colors with Cryostat's colors, but I just reverted back to patternfly's coloring scheme anyways. Probably can just remove this.

File: src/app/app.css (hunk 3 above)
Comment (maxcao13) [reply src/app/app.css:40]: The reason why the overflow is hidden is because if you open the tour you are able to start scrolling past the content a little. I could not really pinpoint why other than it may be a problem with `react-joyride`, but putting this attribute solves the problem.

Comment (maxcao13) [reply src/app/app.css:40]: image

As for the side bar being collapse, this is what happens if the css rule is turned off, but once again, I'm not really sure why. I think it's because `react-joyride` sets styles on the highlighted elements to some overflow attribute. I can actually see it here, inlined: 
image

**Comment by maxcao13** [comment]
No other planned quickstarts for now. Categories of quickstarts will be another time as well, since there isn't enough to justify it yet.

File: src/app/Joyride/CryostatJoyride.tsx (hunk 8 above)
Comment (tthvo) [reply src/app/Joyride/CryostatJoyride.tsx:84]: For me, just resize the window till the nav close, and start guided tour :))

File: src/app/app.css (hunk 3 above)
Comment (maxcao13) [reply src/app/app.css:40]: > Any reason to set overflow as hidden here?

I realize that whitespace is because of this issue: https://github.com/cryostatio/cryostat-web/issues/863

**Comment by andrewazores** [comment]
image

This text isn't quite accurate - active recordings in the `STOPPED` state are not the same thing as archived recordings, but this text seems to suggest that they are. I think it should explain that archived recordings are ones where the data has been copied to Cryostat's storage volume, whereas active recordings exist only in the target application.

**Comment by andrewazores** [comment]
image

This probably needs to have a step telling the user that they need to select the checkbox corresponding to the recording, otherwise the `STOP` button is still disabled if they just follow this directly.

**Comment by andrewazores** [comment]
image

Probably worth calling out that the archives also retain data after the target JVM where the data came from exits.

**Comment by andrewazores** [comment]
Might it make sense to put a prerequisite that the user should complete the "Start a Recording" quickstart before they do "Automated Rules"? Learning how to automate recording creation and archival will make more sense if they already have the context from doing those things manually.

File: src/app/BreadcrumbPage/BreadcrumbPage.tsx [hunk 12]
```diff
@@ -52,8 +61,8 @@ export interface BreadcrumbTrail {
 
 export const BreadcrumbPage: React.FC<BreadcrumbPageProps> = (props) => {
   return (
-    <>
-      <PageSection>
+    <PageGroup>
+      <PageBreadcrumb>
```
Comment (tthvo) [inline src/app/BreadcrumbPage/BreadcrumbPage.tsx:65]: Sorry, can we revert these changes? Seems like this wrapper adds a white background which is quite weird :))

Screenshot from 2023-03-28 16-02-32

File: src/app/app.css (hunk 3 above)
Comment (tthvo) [reply src/app/app.css:40]: > I realize that whitespace is because of this issue

Yehhh just the skip-to-content is out of place. So we should not set hidden here as overflow content on body should be scrollable or else things appear stuck.

Comment (tthvo) [reply src/app/app.css:40]: > As for the side bar being collapse, this is what happens if the css rule is turned off. I think it's because react-joyride sets styles on the highlighted elements to some overflow attribute. I can actually see it here, inlined:

Ahh i see thanks!

File: src/app/AppLayout/AppLayout.tsx [hunk 13]
```diff
@@ -220,27 +233,38 @@ const AppLayout: React.FC<AppLayoutProps> = ({ children }) => {
   const dismissSslErrorModal = React.useCallback(() => setShowSslErrorModal(false), [setShowSslErrorModal]);
 
   const onNavToggleMobile = React.useCallback(() => {
```
Comment (tthvo) [inline src/app/AppLayout/AppLayout.tsx:235]: Remove unused vars here?

File: src/app/AppLayout/AppLayout.tsx [hunk 14]
```diff
@@ -220,27 +233,38 @@ const AppLayout: React.FC<AppLayoutProps> = ({ children }) => {
...
+  }, [setIsNavOpen, joyState, setJoyState]);
 
+  // prevent page resize to close nav during tour
   const onPageResize = React.useCallback(
     (props: { mobileView: boolean; windowSize: number }) => {
-      setIsMobileView(props.mobileView);
+      if (joyState.run === false) {
+        setIsMobileView(props.mobileView);
+        if (props.mobileView) {
+          setIsNavOpen(false);
```
Comment (tthvo) [inline src/app/AppLayout/AppLayout.tsx:254]: Seems like if u resize to smaller width, then resize back. The menu remains closed. Is this the desired behaviour now?

Comment (maxcao13) [reply src/app/AppLayout/AppLayout.tsx:254]: Oh, my bad, I forgot what the old implementation was, I will revert that.

**Review by tthvo — COMMENTED** [review]
Implementations and contents look good to me! Leaving to Andrew for finalizing contents :))

**Comment by andrewazores** [comment]
One last little nitpick: please remove mentions of the word "upstream", and wherever this is referring to some URL, please ensure that URL is provided by the `build.json` file rather than hardcoded into the sources.

File: src/app/Joyride/CryostatJoyride.tsx [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+            </p>
+          ),
+          target: '*[data-tour-id="application-launcher"]',
+        },
+        {
+          title: 'You’re ready to go!',
+          content: (
+            <p>
+              Stay up-to-date with everything Cryostat on our{' '}
+              <a target="_blank" href={`${build.homePageUrl}/blog/`} rel="noreferrer">
```
Comment (andrewazores) [inline src/app/Joyride/CryostatJoyride.tsx:201]: I think this whole URL should be extracted to the `build.json` file, and same for any other cases (like a few lines lower) where a path is resolved against one of the keys from that file. It may be more than just the base of the URL that needs to be patched between an upstream and downstream build, ie. it may not be the case that a downstream patched version uses a URL with the path ending in `/blog` here.

Comment (maxcao13) [reply src/app/Joyride/CryostatJoyride.tsx:201]: I added a key called 
```json
  "documentationUrl": "https://cryostat.io/guides",
```
 Something like `{build.documentationUrl}/#create-an-automated-rule` should be extracted to `build.json` as well like `build.createAutomatedRuleGuideUrl` ?

Comment (andrewazores) [reply src/app/Joyride/CryostatJoyride.tsx:201]: For now let's just go with that. Again, it could be that we change the upstream website structure or there is a downstream patched build that needs a different structure, and maybe the guide section isn't a page anchor but a different on the server, for example.

If that ends up being a significant annoyance/maintenance burden with a lot of instances of copy-pasted relative URLs like that then we can look for a specific solution for that problem.
